File: Arrays/first-missing-positive/first-missing-positive.ts
// https://leetcode.com/problems/first-missing-positive/

function firstMissingPositive(nums: number[]): number {
  const n = nums.length;
  let i = 0;
  while (i < n) {
    let index = nums[i] - 1;
    if (index >= 0 && index < n && nums[index] != nums[i]) {
      let temp = nums[index];
      nums[index] = nums[i];
      nums[i] = temp;
    } else {
      i++;
    }
  }

  for (let i = 0; i < n; i++) {
    if (nums[i] != i + 1) {
      return i + 1;
    }
  }
  return n + 1;
}

File: Arrays/first-missing-positive/first-missing-positive.go
// https://leetcode.com/problems/first-missing-positive/

package main
func firstMissingPositive(nums []int) int {
    n := len(nums)
    i := 0;
    for i < n {
        index := nums[i] - 1
        if index >= 0 && index < n && nums[index] != nums[i] {
            nums[index], nums[i] = nums[i], nums[index]
        } else {
            i++
        }
    }
    for i := 0; i < n; i++ {
        if nums[i] != i+1 {
            return i + 1
        }
    }
    return n + 1
}